discord/guild.go

```diff
@@ -366,6 +367,104 @@ func (guild *GuildState) handleReactionGameStartAdd(s *discordgo.Session, m *dis
 
 }
 
+func (guild *GuildState) handleReactionPrivateUserMessage(s *discordgo.Session, m *discordgo.MessageReactionAdd) {
+	if !guild.PrivateStateMsg.Exists() || !guild.PrivateStateMsg.IsReactionTo(m) {
+		return
+	}
+
+	var idUsernameMap = guild.PrivateStateMsg.idUsernameMap;
```

Comment:
These *shouldn't* cause race conditions under the typical use case (these are DMs, correct?), but, it might be good practice to enforce locking/unlocking regardless. If there is no contention for the lock, then the performance hit is practically 0, but if a user accidentally clicks 2 reactions on accident, for example, then it guarantees that the state doesn't get borked via a race condition problem

Reply:
Would that look similar to what I showed above in another comment?:

```
func (psm *PrivateStateMessage) getIDUsernameMap() map[string]string{
	psm.lock.RLock()
	defer psm.lock.RUnlock()

	return psm.idUsernameMap
}
```

Also, these messages are in a specified text channel of the server.

Reply:
That would work, or even to take it to another level, you could have a function that also fetches values from the map *for you*
example:
```
func (psm *PrivateStateMessage) getIDUsername(id string) string{
	psm.lock.RLock()
	defer psm.lock.RUnlock()

	return psm.idUsernameMap[id]
}
```
(this example ignores properly handling if the map doesn't have the value).

But the advantage here is that, like I mentioned above for the "constructor", *we don't have to worry about locks elsewhere in the code base*. Everything about the locking/thread-safety of the map is completely abstracted away; we just call `getIdUsername` blindly, and don't worry about handling the locks ourselves (despite it still being completely thread-safe)

Reply:
I understand what you're saying. I've tried to implement just now, but I run into a problem when I need a map for these lines of code:

```
for uId, uName := range idUsernameMap {


		//log.Print("Showing printedUsers again...")
		var shownUsername = false;
		for _, username := range printedUsers {
			//log.Print("user: " + username);
			if username == uId {
				shownUsername = true;
			}
		}

		log.Print("uID:" + uId)

		if (!shownUsername) {
			log.Print("Haven't shown this user: " + uName);
			log.Print("users id: " + uId);
			userId = uId;
			userName = uName;
			//username = uName;
			break;
		}
	}
```

Reply:
Does this implementation work?

Inside of gameStateMessage.go:

```
func (psm *PrivateStateMessage) getIDUsername(id string) string{
	psm.lock.RLock()
	defer psm.lock.RUnlock()

	return psm.idUsernameMap[id]
}

func (psm *PrivateStateMessage) getStoredIds() []string{
	psm.lock.RLock()
	defer psm.lock.RUnlock()

	keys := make([]string, 0, len(psm.idUsernameMap))
	for k := range psm.idUsernameMap {
		keys = append(keys, k)
	}

	return keys
}
```

Inside of guild.go:
```
userId = "";
	var userName string;
	for _, uId := range guild.PrivateStateMsg.getStoredIds() {

		var uName = guild.PrivateStateMsg.getIDUsername(uId);


		//log.Print("Showing printedUsers again...")
		var shownUsername = false;
		for _, username := range printedUsers {
			//log.Print("user: " + username);
			if username == uId {
				shownUsername = true;
			}
		}

		log.Print("uID:" + uId)

		if (!shownUsername) {
			log.Print("Haven't shown this user: " + uName);
			log.Print("users id: " + uId);
			userId = uId;
			userName = uName;
			//username = uName;
			break;
		}
	}
```